File: 2024 1월 3주차/백준 1707.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
    public static boolean bfs(ArrayList<ArrayList<Integer>> graph, int start, int[] visited) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(start);
        visited[start] = 1;

        while (!queue.isEmpty()) {
            int node = queue.poll();
            for (int adjNode : graph.get(node)) {
                if (visited[adjNode] == 0) {
                    visited[adjNode] = -visited[node];
                    queue.add(adjNode);
                } else if (visited[adjNode] == visited[node]) {
                    return false;
                }
            }
        }
        return true;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int K = scanner.nextInt();

        for (int k = 0; k < K; k++) {
            int V = scanner.nextInt();
            int E = scanner.nextInt();

            ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
            for (int i = 0; i <= V; i++) {
                graph.add(new ArrayList<>());
            }

            int[] visited = new int[V + 1];
            for (int i = 0; i < E; i++) {
                int u = scanner.nextInt();
                int v = scanner.nextInt();
                graph.get(u).add(v);
                graph.get(v).add(u);
            }

            boolean isBipartite = true;
            for (int i = 1; i <= V; i++) {
                if (visited[i] == 0 && !bfs(graph, i, visited)) {
                    isBipartite = false;
                    break;
                }
            }
            System.out.println(isBipartite ? "YES" : "NO");
        }
        scanner.close();
    }
}

File: 2024 1월 3주차/백준 1707.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct Node {
    int vertex;
    struct Node* next;
} Node;

Node* createNode(int v) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->vertex = v;
    newNode->next = NULL;
    return newNode;
}

void addEdge(Node** graph, int u, int v) {
    Node* newNode = createNode(v);
    newNode->next = graph[u];
    graph[u] = newNode;

    newNode = createNode(u);
    newNode->next = graph[v];
    graph[v] = newNode;
}

bool bfs(Node** graph, int start, int* visited, int V) {
    int* queue = (int*)malloc((V + 1) * sizeof(int));
    int front = 0, rear = 0;
    visited[start] = 1;
    queue[rear++] = start;

    while (front < rear) {
        int node = queue[front++];
        Node* temp = graph[node];
        while (temp) {
            int adjVertex = temp->vertex;
            if (visited[adjVertex] == 0) {
                visited[adjVertex] = -visited[node];
                queue[rear++] = adjVertex;
            } else if (visited[adjVertex] == visited[node]) {
                free(queue);
                return false;
            }
            temp = temp->next;
        }
    }
    free(queue);
    return true;
}

int main() {
    int K;
    scanf("%d", &K);
    for (int k = 0; k < K; k++) {
        int V, E;
        scanf("%d %d", &V, &E);
        Node** graph = (Node**)malloc((V + 1) * sizeof(Node*));
        int* visited = (int*)malloc((V + 1) * sizeof(int));

        for (int i = 1; i <= V; i++) {
            graph[i] = NULL;
            visited[i] = 0;
        }

        for (int i = 0; i < E; i++) {
            int u, v;
            scanf("%d %d", &u, &v);
            addEdge(graph, u, v);
        }

        bool isBipartite = true;
        for (int i = 1; i <= V; i++) {
            if (visited[i] == 0 && !bfs(graph, i, visited, V)) {
                isBipartite = false;
                break;
            }
        }
        printf("%s\n", isBipartite ? "YES" : "NO");

        for (int i = 1; i <= V; i++) {
            Node* temp;
            while (graph[i]) {
                temp = graph[i];
                graph[i] = graph[i]->next;
                free(temp);
            }
        }
        free(graph);
        free(visited);
    }
    return 0;
}